chain/syncer.go

```diff
@@ -421,13 +345,39 @@ func (syncer *Syncer) HandleNewTipset(ctx context.Context, tsKey types.TipSetKey
 		return nil
 	}
 
-	// Walk the chain given by the input blocks back to a known tipset in
-	// the store. This is the only code that may go to the network to
-	// resolve cids to blocks.
-	chain, err := syncer.collectChain(ctx, tsKey)
-	if err != nil {
-		return err
+	haltOnState := func(t types.TipSetKey) bool {
+		if syncer.chainStore.HasTipSetAndState(ctx, t.String()) {
+			return true
+		}
+		return false
+	}
+
+	var chain []types.TipSet
+	switch syncer.syncMode {
+	case Syncing:
+		chain, err = syncer.fetcher.FetchTipSets(ctx, tsKey, haltOnState)
+		if err != nil {
+			return err
+		}
+		chain = reverse(chain)
+	case CaughtUp:
+		// we cancel the fetch during caught up mode after a single blocktime.
+		fetchCtx, cancel := context.WithTimeout(ctx, blkWaitTime)
+		defer cancel()
+		chain, err = syncer.fetcher.FetchTipSets(fetchCtx, tsKey, haltOnState)
```

Comment:
Blocking: it's more complicated than that. In this state, we aim to avoid fetching a giant chain that turns out not to join with ours, so the finality check has to come *before* the big fetch.

I think this should:
- fetch one tipset
- check whether it exceeds the finality limit
- if not, fetch the rest